### PR TITLE
Feature/fspace repartition ecmwf

### PR DESCRIPTION
**Pull request by odlomax**
## Description

Added functionality to repartition fields  between two StructuredColumns function spaces with the same grid, but different partitioners. All Atlas tests pass in my gcc  8.1 release build.

## Dependencies

None. Direct calls to MPI library have been removed.

### Build

Updated from C++11 to C++14.

### atlas/src/atlas/repartition

_Repartition.h; Repartition.cc_
Base class for repartition object.

_detail/RepartitionImpl.h; detail/RepartitionImpl.cc_
Abstract base class for repartition implementation.

_detail/StructuredColumnsToStructuredColumns.h; detail/StructuredColumnsToStructuredColumns.cc_
Concrete implementation for StructuredColumns function spaces.

_detail/RepartitionImplFactory.h; detail/RepartitionImplFactory.cc_
Factory class for RepartitionImpl objects. Currently only one concrete implementation to choose from.

_detail/RepartitionUtils.h_
Utilities for run-time type, size and cast checking.

### atlas/src/tests/repartition

_test_repartition.cc_
Series of tests which run repartitioner on various grid/partitioner combinations. Also output some gmsh mesh and field files to corresponding sub directory in build.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](http://claassistant.ecmwf.int:80/ecmwf/atlas?pullRequest=62) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](http://claassistant.ecmwf.int:80/ecmwf/atlas?pullRequest=62) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://claassistant.ecmwf.int:80/check/ecmwf/atlas?pullRequest=62) it.</sub>

**Comment by wdeconinck** [comment]
Thank you @odlomax I will review asap. I have in the mean time ran it through our in-house CI, and noticed the following:

```
In file included from atlas-bundle/source/atlas/src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc:16:0:
15-Mar-2021 20:57:52 | atlas-bundle/source/atlas/src/atlas/repartition/detail/RepartitionUtils.h:10:17: fatal error: mpi.h: No such file or directory
15-Mar-2021 20:57:52 | #include <mpi.h>
15-Mar-2021 20:57:52 | ^
15-Mar-2021 20:57:52 | compilation terminated.
```
So here is a reference to `mpi` directly still.  You can compile eckit with `-DENABLE_MPI=OFF` for your test to make sure that you are not using MPI directly.

On our Cray cce/8.7 I have:
```
atlas: symbol lookup error: /opt/cray/lib64/libmpichcxx_cray.so.3: undefined symbol: __eh_curr_region
```
Usually we don't use the cxx mpi API. This issue will also disappear when all MPI references have gone.

Then additionally I noticed that GCC 5 compiler tested with Travis fails.
I agree that this is a quite old compiler, but for instance Intel 17 compiler and our oldest relevant Cray compiler (cce/8.5) are still required for deployment and uses GNU 4.x!! headers.

The least headache would be to try to find a C++11 workaround for now unfortunately.

**Comment by odlomax** [comment]
> Thank you @odlomax I will review asap. I have in the mean time ran it through our in-house CI, and noticed the following:
> 
> ```
> In file included from atlas-bundle/source/atlas/src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc:16:0:
> 15-Mar-2021 20:57:52 | atlas-bundle/source/atlas/src/atlas/repartition/detail/RepartitionUtils.h:10:17: fatal error: mpi.h: No such file or directory
> 15-Mar-2021 20:57:52 | #include <mpi.h>
> 15-Mar-2021 20:57:52 | ^
> 15-Mar-2021 20:57:52 | compilation terminated.
> ```
> 
> So here is a reference to `mpi` directly still. You can compile eckit with `-DENABLE_MPI=OFF` for your test to make sure that you are not using MPI directly.
> 
> On our Cray cce/8.7 I have:
> 
> ```
> atlas: symbol lookup error: /opt/cray/lib64/libmpichcxx_cray.so.3: undefined symbol: __eh_curr_region
> ```
> 
> Usually we don't use the cxx mpi API. This issue will also disappear when all MPI references have gone.
> 
> I also found following with GCC 6.3 which seems to be the same how Travis fails with GCC 5:
> 
> ```
> 15-Mar-2021 21:32:58 | [1785/2419] Building CXX object atlas/src/atlas/CMakeFiles/atlas.dir/repartition/detail/StructuredColumnsToStructuredColumns.cc.o
> -- | --
> 15-Mar-2021 21:32:58 | FAILED: atlas/src/atlas/CMakeFiles/atlas.dir/repartition/detail/StructuredColumnsToStructuredColumns.cc.o
> 15-Mar-2021 21:32:58 | /usr/local/apps/gcc/6.3.0/bin/g++  -Datlas_EXPORTS -Iatlas/module -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/atlas/src/atlas -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/atlas/src -Iatlas/src -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/fckit/src -Ifckit/src -Ifckit/module -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/fckit/src/fckit -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/transi/src/transi -Itransi/module -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/transi/src -Itransi/src -Ieckit/src -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/eckit/src -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/eckit/src/eckit -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/eckit/src/eckit/geometry -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/eckit/src/eckit/linalg -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/eckit/src/eckit/mpi -I/tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/eckit/src/eckit/option -I/usr/local/apps/proj/6.1.1/include -I/usr/local/apps/cgal/4.10/GNU/6.3.0/include -I/usr/local/apps/boost/1.61.0/GNU/6.3.0/include -I/usr/local/apps/intel/18.0.1/mkl/include/fftw -I/usr/local/apps/openmpi/2.1.3/GNU/6.3.0/include -I/usr/local/apps/eigen/3.3.4/include/eigen3 -pipe -O0 -g -fPIC   -fopenmp -std=gnu++14 -MD -MT atlas/src/atlas/CMakeFiles/atlas.dir/repartition/detail/StructuredColumnsToStructuredColumns.cc.o -MF atlas/src/atlas/CMakeFiles/atlas.dir/repartition/detail/StructuredColumnsToStructuredColumns.cc.o.d -o atlas/src/atlas/CMakeFiles/atlas.dir/repartition/detail/StructuredColumnsToStructuredColumns.cc.o -c /tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/atlas/src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc
> 15-Mar-2021 21:32:58 | /tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/atlas/src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc: In instantiation of ‘atlas::repartition::detail::StructuredColumnsToStructuredColumns::execute(const atlas::FieldSet&, atlas::FieldSet&) const::<lambda(const auto:9&)> mutable [with auto:9 = atlas::Field]’:
> 15-Mar-2021 21:32:58 | /usr/local/apps/gcc/6.3.0/lib64/gcc/x86_64-suse-linux/6.3.0/include/c++/bits/stl_algo.h:3769:5:   required from ‘_Funct std::for_each(_IIter, _IIter, _Funct) [with _IIter = __gnu_cxx::__normal_iterator<const atlas::Field*, std::vector<atlas::Field> >; _Funct = atlas::repartition::detail::StructuredColumnsToStructuredColumns::execute(const atlas::FieldSet&, atlas::FieldSet&) const::<lambda(const auto:9&)>]’
> 15-Mar-2021 21:32:58 | /tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/atlas/src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc:197:10:   required from here
> 15-Mar-2021 21:32:58 | /tmp/bamboo-agent-okolnir-b/xml-data/build-dir/ATLAS-BUNDLEGH-JOB1/source/atlas/src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc:195:11: error: cannot call member function ‘virtual void atlas::repartition::detail::StructuredColumnsToStructuredColumns::execute(const atlas::Field&, atlas::Field&) const’ without object
> 15-Mar-2021 21:32:58 | execute(sourceField, *targetFieldSetIt++);
> 15-Mar-2021 21:32:58 | ^~~~~~~
> ```
> 
> This could be due to C++14 not being fully supported in these compilers. I hope you can find a workaround for now, until we can completely disregard these older compilers.

Whoops! The `#include <mpi.h>` did not belong there! That'll teach me to push before I grep. I'll rewrite some bits back to C++11 standard. Shouldn't take long,

**Comment by odlomax** [comment]
Refactored to the code to C++11. I've also added a few different data type tests for good measure. Couldn't seem to create fields with type `<unsigned long>`. At least `<double>`, `<float>`, `<int>` and `<long>` all work.

**Comment by odlomax** [comment]
@wdeconinck  Looks like Travis CI fell over trying to apt-get gcc. Would you mind giving it a nudge?

File: src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.h [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  }
+}
+
+namespace atlas {
+  namespace repartition {
+    namespace detail {
+
+      // Forward declarations.
+      class StructuredColumnsToStructuredColumns;
+      class FuncSpaceRange;
```
Comment (wdeconinck) [inline src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.h:34]: Please rename `FuncSpaceRange` to `detail::StructuredIndexRange` to be more unique and descriptive to what it applies

**Review by wdeconinck — CHANGES_REQUESTED** [review]
Hi @odlomax ,  Very high quality code!
Please address a few minor cosmetic changes about naming.

Did you check if redistribution works with different values for halo for source and target?

File: src/atlas/repartition/Repartition.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    /// \details  Transfers source field set to target field set via multiple
+    ///           invocations of execute(sourceField, targetField).
+    ///
+    /// \param[in]  sourceFieldSet  input field set.
+    /// \param[out] targetFieldSet  output field set.
+    void execute(
+      const FieldSet& sourceFieldSet, FieldSet& targetFieldSet) const;
+
+    /// \brief  Get reference to source function space.
+    FunctionSpace& getSourceFunctionSpace();
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/atlas/repartition/Repartition.h]: Done.

File: src/atlas/repartition/Repartition.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+#include "atlas/util/ObjectHandle.h"
+
+namespace atlas {
+  class Field;
+  class FieldSet;
+  class FunctionSpace;
+}
+
+namespace atlas {
+
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/atlas/repartition/Repartition.h]: Done. Much better name, especially given that the actual partitioning is handled by another class.

p.s. apologies if "redistributioner" works its way into the comments somewhere!

File: src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+  namespace repartition {
+    namespace detail {
+
+      // Anonymous name space for helper functions.
+      namespace {
+
+        // Transform a vector element-by-element with a functor.
+        template <typename outType, typename inType, typename functorType>
+        std::vector<outType> transformVector(
+          const std::vector<inType>& inVector, functorType functor) {
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc]: Done. I believe copies are safe in this instance, but const reference is probably better practice.

File: src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+          std::transform(inVector.begin(), inVector.end(),
+            std::back_inserter(outVector), functor);
+
+          return outVector;
+        }
+
+        // Visit each index in a FuncSpaceRangeVector.
+        template <typename functorType>
+        void forEachIndex(const FuncSpaceRangeVector& ranges,
+          functorType functor) {
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc]: Done.

File: src/atlas/repartition/detail/RepartitionUtils.h [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+        // Check if cast failed.
+        if (!(functionSpacePtr->cast<FunctionSpaceType>())) {
+           throw eckit::BadCast("Cannot cast " + varName + " to " +
+           typeid(FunctionSpaceType).name() , location);
+        }
+
+        return;
+      }
+      #define TRY_CAST(a, b) tryCast<a>(b, #b, Here())
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/atlas/repartition/detail/RepartitionUtils.h]: Good call. Macros in a .h file could cause all kinds of carnage. I've moved them to the .cc.

It's my first time using macros. They're seem pretty useful, although I'm sure they make Bjarne Stroustrup cry.

File: src/atlas/repartition/detail/RepartitionImplFactory.cc [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            sourceFunctionSpace->type() + " and target functionspace " +
+            targetFunctionSpace->type(), Here());
+          return nullptr;
+
+        }
+      }
+    }
+  }
+}
+
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/atlas/repartition/detail/RepartitionImplFactory.cc]: Thanks. I'm going to have to revisit this in the future to implement a more general version of this algorithm. That'll be a good time to use self-registration.

File: src/atlas/repartition/detail/RepartitionImpl.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        FunctionSpace& getSourceFunctionSpace();
+
+        /// \brief  Get const reference to source function space.
+        const FunctionSpace& getSourceFunctionSpace() const;
+
+        /// \brief  Get reference to target function space.
+        FunctionSpace& getTargetFunctionSpace();
+
+        /// \brief  Get const reference to target function space.
+        const FunctionSpace& getTargetFunctionSpace() const;
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/atlas/repartition/detail/RepartitionImpl.h]: Done.

File: src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+
+          const auto counts = transformVector<int>(intersections,
+            [&](const FuncSpaceRange& intersection) {
+
+            return static_cast<int>(intersection.getElemCount() * levels);
+          });
+
+          auto displacements = std::vector<int>{0};
+          std::partial_sum(counts.cbegin(), counts.cend() - 1,
+            std::back_inserter(displacements));
```
[the remark this answers is not in the document]
Comment (odlomax) [reply src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.cc]: Thanks! I enjoy the sport of using STL _aggressively_.

File: src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.h (hunk 1 above)
Comment (odlomax) [reply src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.h]: Good point. I originally called it `IndexRange`, then I learned that this was a thing elsewhere. In my panic, I renamed it something rubbish! This is a better name.

Comment (odlomax) [reply src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.h]: > Hi @odlomax , Very high quality code!
> Please address a few minor cosmetic changes about naming.
> 
> Did you check if redistribution works with different values for halo for source and target?

Thanks @wdeconinck  for your kind words! I've addressed your comments and added a couple tests with mixed halo sizes.

**Review by wdeconinck — APPROVED** [review]

**Comment by wdeconinck** [comment]
This is merged manually in develop now. Thanks @odlomax !

**Pull request closed without merging**